Loop/Models/Deeplink.swift

```diff
@@ -7,18 +7,69 @@
 //
 
 import Foundation
+import LoopAlgorithm
 
-enum Deeplink: String, CaseIterable {
-    case carbEntry = "carb-entry"
-    case bolus = "manual-bolus"
-    case preMeal = "pre-meal-preset"
-    case customPresets = "custom-presets"
+enum Deeplink: Hashable {
+    
+    struct AppSource: Hashable {
+        let name: String
+        let bundleId: String?
+    }
+    
+    enum Host: String, CaseIterable {
+        case carbEntry = "carb-entry"
+        case bolus = "manual-bolus"
+        case preMeal = "pre-meal-preset"
+        case customPresets = "custom-presets"
+    }
+    
+    enum CarbEntryLink: Hashable {
+        case carbEntryDetected(value: LoopQuantity, source: AppSource?)
+    }
+    
+    case carbEntry(CarbEntryLink?)
+    
+    case bolus
+    case preMeal
+    case customPresets
+    
+    var host: Host {
+        switch self {
+        case .carbEntry: .carbEntry
+        case .bolus: .bolus
+        case .preMeal: .preMeal
+        case .customPresets: .customPresets
+        }
+    }
     
     init?(url: URL?) {
-        guard let url, let host = url.host, let deeplink = Deeplink.allCases.first(where: { $0.rawValue == host }) else {
+        guard let url, let host = url.host, let deeplinkHost = Deeplink.Host.allCases.first(where: { $0.rawValue == host }) else {
             return nil
         }
+
+        let components = URLComponents(url: url, resolvingAgainstBaseURL: true)
         
-        self = deeplink
+        switch deeplinkHost {
+        case .carbEntry:
+            if let value = components?.queryItems?.first(where: { $0.name == "value" })?.value, let doubleValue = Double(value) {
+                let sourceBundleId = components?.queryItems?.first(where: { $0.name == "sourceBundleId" })?.value
+                let sourceName = components?.queryItems?.first(where: { $0.name == "sourceName" })?.value
```

Comment:
I'm wondering about possible misuse of this; any app could send a spoofed name to make the carb entry look it it came from a different app.

Reply:
Or is sourceName coming from iOS here?